### PR TITLE
Django 1.8 migration in 29 commits

### PR DESCRIPTION
**Pull request by vladimir-mencl-eresearch**
Hi Zenon,

This is the Django 1.8 migration - I believe it is complete.

I have run a fairly detailed test coverage (all URLs, aiming for all different actions, also invoking all management commands) and as far as I can tell, everything is working.

I have added documentation for migration.  It will need users to first push forward to the latest 1.4 release (including the InstServer M2M change), but after that, the instructions are straightforward.  (We will need to tag the tip of the next branch and make it a stable reference point prior to this merge).

Please let me know if you're happy to merge this (or any concerns you may have).

Cheers,
Vlad

### CODE REVIEW

**Comment by zmousm** [comment]
Thanks for the hard work Vladimir! Please allow us a few days to process this.

**Comment by vladimir-mencl-eresearch** [comment]
Hi Zenon,

Just wondering whether you've had a chance to reviews this....

Cheers,
Vlad

PS:  And, btw, I've just pushed in one more commit with a minor change - bumping python-social-auth from 0.2.10 up to 0.2.14 - just to get bugfixes.

**Comment by vladimir-mencl-eresearch** [comment]
Hi Zenon,

FYI, I have also developed the feature to make the list of login methods configurable (as discussed in #9).  I have it so far in the develop branch at REANNZ/djnro - and I've used the django18 branch as the starting point.

Once you merge the django18 branch, I'd create a pull request for this feature.  In the meantime, you can have a look at the login methods feature at https://github.com/REANNZ/djnro/compare/django18...REANNZ:develop

Any feedback (either on the login methods feature or the Django 1.8 migration) would be greatly appreciated.

Cheers,
Vlad

**Comment by zmousm** [comment]
Hi Vladimir, we have a first round of comments that I will be posting shortly... It's a lot of files though and we've been looking at individual commits so I'm just figuring out how to move comments to PR diff context.

**Comment by vladimir-mencl-eresearch** [comment]
Hi Zenon,

I hope I answered all of your questions - would you be happy to merge this now?

Cheers,
Vlad

**Comment by zmousm** [comment]
Hi again Vladimir,

beyond this first set of minor issues that you have already mostly answered, we want to do some more tests for the migration/upgrade process to Django 1.8, so as to be able to better understand what you have documented in REANNZ@8c8842e. We will thus need a few more days for that.

Regards,
Z.

**Comment by vladimir-mencl-eresearch** [comment]
Hi Zenon,

I have updated commit messages for:
- Django 1.8: update links to RegistrationProfile
  (fixed the wrong order of old/new links)
- Dj1.8: CustomUM: accounts.User: properly swappable
  (added explanation as per our discussion on the commit itself)
- Django1.8: Custom User Model: port auth migrations
  (added explanation why we need separate migrations for changes in the Auth model)

I understand you still need to review the migration instructions.  I have successfully tested them on my own DEV systems.

Btw, I also have a script that does that in an automated way (as part of our Dockerized deployment) - I'd be happy to contribute that script too if you'd find it useful:

https://github.com/REANNZ/etcbd/blob/master/environment/djnro/content/etc/startup.d/50-dj18upgrade.sh

Please let me know if you have any further questions - and I look forward to hearing from you once you are able to merge this PR.

Cheers,
Vlad

**Comment by vladimir-mencl-eresearch** [comment]
Hi Zenon,

FYI, I have pushed in a minor change to the upgrade step: keeping the door open to add new migrations to the edumanage models without having instructions in place that would tell the user to ignore (fake) any such new migrations.  The `fake` step is now limited to 0001_initial and any subsequent migrations should now get properly applied.

Cheers,
Vlad

**Comment by zmousm** [comment]
Vladimir, I have a couple more comments to add for the migration instructions. Stay tuned!

**Comment by vladimir-mencl-eresearch** [comment]
Thanks, keen to hear from you.

Cheers,
Vlad

**Comment by zmousm** [comment]
With regard to the migration script, which could be considered complementary to the upgrade/migration instructions.

First of all I think it can be quite useful, and it looks to me that it's not actually tied to use with Docker. However it's obviously not generic enough (e.g. blindly adding the PostgreSQL constraint).

I don't think we have any particular experience with tools for managed installations/upgrades/etc. specific to Django projects, yet in our ops environment folks seem to be fond of [Fabric](http://www.fabfile.org/). I can see there is a [comparison matrix](https://www.djangopackages.com/grids/g/deployment/) of tools for such a job.

I think we could get a better perspective how one may handle such tasks and perhaps revisit this topic (and maybe update/simplify instructions) after merging this PR.

**Comment by zmousm** [comment]
Vladimir, I can now say that we are done with all our comments and are ready to merge this PR, mainly pending your last answers/final touches for the upgrade instructions. We still want to do more testing with real usage, but we are confident we can do this after merging and report+fix any issues that may arise as necessary.

Yay! Thanks for bearing with us :+1:

**Comment by vladimir-mencl-eresearch** [comment]
Hi Zenon,

Thanks for the patience too!

I've just finished replying to your latest comments - I look forward to hearing from you back.

Cheers,
Vlad

**Comment by vladimir-mencl-eresearch** [comment]
Hi Zenon,

Just wondering there there are any pending issues to resolve...

Cheers,
Vlad

File: accounts/models.py [hunk 1]
```diff
@@ -1,10 +1,33 @@
...
+User._meta.get_field('username').max_length = 255
+# try also updating the help_text
+try:
+    help_text = User._meta.get_field('username').help_text._proxy____args[0]
+    if help_text != None:
+	User._meta.get_field('username').help_text = _(re.sub("[0-9]+ characters", "255 characters", help_text))
+except:
+    pass
+
+admin.site.register(User, UserAdmin)
```
Comment (zmousm) [inline accounts/models.py]: While trying something irrelevant, namely checking if `django-registration` works with Django 1.7 (not that we expected djnro to work as is in such a setup), we noticed `./manage.py shell` borks apparently due to this line with an 'AppRegistryNotReady' exception:

``` python
Traceback (most recent call last):
  File "./manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 385, in execute_from_command_line
    utility.execute()
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 354, in execute
    django.setup()
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/__init__.py", line 21, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/apps/registry.py", line 108, in populate
    app_config.import_models(all_models)
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/apps/config.py", line 202, in import_models
    self.models_module = import_module(models_module_name)
  File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/srv/djnro/accounts/models.py", line 27, in <module>
    admin.site.register(User, UserAdmin)
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/contrib/admin/sites.py", line 101, in register
    system_check_errors.extend(admin_class.check(model))
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/contrib/admin/options.py", line 153, in check
    return cls.checks_class().check(cls, model, **kwargs)
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/contrib/admin/checks.py", line 494, in check
    errors.extend(self._check_list_filter(cls, model))
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/contrib/admin/checks.py", line 665, in _check_list_filter
    for index, item in enumerate(cls.list_filter)
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/contrib/admin/checks.py", line 710, in _check_list_filter_item
    get_fields_from_path(model, field)
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/contrib/admin/utils.py", line 457, in get_fields_from_path
    fields.append(parent._meta.get_field_by_name(piece)[0])
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/db/models/options.py", line 416, in get_field_by_name
    cache = self.init_name_map()
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/db/models/options.py", line 445, in init_name_map
    for f, model in self.get_all_related_m2m_objects_with_model():
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/db/models/options.py", line 563, in get_all_related_m2m_objects_with_model
    cache = self._fill_related_many_to_many_cache()
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/db/models/options.py", line 577, in _fill_related_many_to_many_cache
    for klass in self.apps.get_models():
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/utils/lru_cache.py", line 101, in wrapper
    result = user_function(*args, **kwds)
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/apps/registry.py", line 168, in get_models
    self.check_models_ready()
  File "/tmp/virtualenv/djnro17/local/lib/python2.7/site-packages/django/apps/registry.py", line 131, in check_models_ready
    raise AppRegistryNotReady("Models aren't loaded yet.")
```

We think it's a good idea anyway to move this to `accounts/admin.py`, with such a patch:

``` diff
diff --git a/accounts/admin.py b/accounts/admin.py
index ee85a48..79e7d35 100644
--- a/accounts/admin.py
+++ b/accounts/admin.py
@@ -1,8 +1,10 @@
 from django.contrib import admin
-from accounts.models import UserProfile
+from accounts.models import User, UserProfile
+from django.contrib.auth.admin import UserAdmin


 class UserPrAdmin(admin.ModelAdmin):
     list_display = ('user', 'institution')

 admin.site.register(UserProfile, UserPrAdmin)
+admin.site.register(User, UserAdmin)
diff --git a/accounts/models.py b/accounts/models.py
index ff03608..510e93b 100644
--- a/accounts/models.py
+++ b/accounts/models.py
@@ -24,8 +24,6 @@ try:
 except:
     pass

-admin.site.register(User, UserAdmin)
-
 class UserProfile(models.Model):
     user = models.OneToOneField(settings.AUTH_USER_MODEL)
     institution = models.ForeignKey(Institution)
```

Do you agree? If yes, would you mind adding this patch to the PR, or perhaps I can also add this during the merge?

**Comment by zmousm** [comment]
Hey Vladimir,

see my last line note and then we are good to go :) Sorry for not replying earlier, I've gone busy again.

File: accounts/models.py (hunk 1 above)
Comment (vladimir-mencl-eresearch) [inline accounts/models.py]: Hi Zenon,

Good point!  I've checked and the Django auth.User model also does the registration from admin.py.

Done, pushed in in 3e00840.

We should be good to go now!

Cheers,
Vlad

**Comment by zmousm** [comment]
Thanks Vladimir. I will proceed with the merge. I am tempted to squash some (of the 42) commits to clean things up a little. Authorship will be preserved of course. Would you mind terribly?

**Comment by vladimir-mencl-eresearch** [comment]
Hi Zenon, I'll try some squashing.  The commits around Django version are good candidates - and so are about the swappable properties.

**Comment by zmousm** [comment]
I was gonna do it myself, but... great, thanks!

**Comment by vladimir-mencl-eresearch** [comment]
Do you have any other candidates for squashing?

**Comment by zmousm** [comment]
c41c23e, 15a3c1d, ee911b1 maybe, all touching `docs/installation/upgrading-from-1.0.md`.

**Comment by vladimir-mencl-eresearch** [comment]
Hi Zenon,

So, this is my take on `git rebase -i` - reordering and squashing things that can be grouped together.
Well, ended up finding quite a number of iterative revisions.  This should take the number of commits down to 27:

```
pick 66f4a06 Django: start migrating to 1.8.x (BREAKING CHANGE)
squash 4ffaa7e Django 1.8: select Django 1.8.x compatible version
squash ff98cc0 Dj18: bump up minimum Django version to cur 1.8.11
squash 3d537bc Django 1.8: avoid PIP PEP 440 compatible notation
pick 60f7286 Django 1.8.x: use lazy gettext in settings modules
pick 93cef62 Django 1.8.x: replace dj.markup with markdown_deux
squash 4263018 Django 1.8: remove markdown_deux
pick c5dd1ac Django 1.8.x: de-dup django.contrib.staticfiles
pick 4a79af3 Update python-social-auth to 0.2.14
pick 2b93245 Django 1.8: up dj-registration: fix loading models
squash e7267bc Django 1.8: update links to RegistrationProfile
pick 7def67f Django 1.8.x: fix url imports
pick 77f403e Django 1.8.x: update email validation code
squash 47c3ba4 Django 1.8: mark strings for translation
pick 4253e9b Django 1.8.x: explicitly list all fields on forms
pick 47b8435 Django 1.8: rename kwarg mimetype to content_type
pick 1479ad0 Django 1.8.x: quote URL names in templates
pick e6f61f0 Django 1.8.x: Fix accessing user profiles
pick cc20eac Django 1.8: uncomment missing STATIC_ROOT setting
pick 5e68a0e D1.8: replace direct_to_template with TemplateView
pick 9bbcc9d Django1.8: contrib.contenttypes.generic deprecated
squash d0014eb Django1.8: contenttypes.generic deprecated (cont)
pick a4d8d0b Django 1.8: remove South
pick d4baa53 Dj1.8: avoid warnings: remove unused field options
pick 5068d53 Dj1.8: repl longerusername with custom user model
squash 3e00840 Dj18: move CustomUM registration into acc/admin.py
squash f25cc1a Dj1.8: CustomUM: accounts.User: properly swappable
squash 36acce1 Dj1.8: custom user model: update imports
pick 7572c46 Django 1.8: archive South migrations
pick 8752308 Django 1.8: add new generated initial migrations
squash 5056446 Dj1.8: CustomUM: swappable prop also to migrations
pick 0c8cdb3 Django1.8: Custom User Model: port auth migrations
pick f864d05 Django 1.8: locales in management commands
pick eaa14ff Django 1.8 requires non-empty ALLOWED_HOSTS
pick 3fa2d74 Update logging configuration to Django 1.8 default
pick 12a2290 Dj 1.8 logging: filter out DisallowedHost emails
pick 064a146 Django 1.8: add documentation
squash ee911b1 Dj1.8 upgr doc: fake edumange only to 0001_initial
squash 15a3c1d Django 1.8: upgr doc: contenttypes chg before migs
squash c41c23e Dj1.8 upgr doc: rephrase PostgreSQL fix-up instrs
```

Will soon run this and see if all of it can be easily merged...

Cheers,
Vlad

**Comment by zmousm** [comment]
Alright, but I think that for some sequences of squashed commits, you want to keep the first one. More specifically:

```
pick 66f4a06 Django: start migrating to 1.8.x (BREAKING CHANGE)
squash->pick 4ffaa7e Django 1.8: select Django 1.8.x compatible version
squash ff98cc0 Dj18: bump up minimum Django version to cur 1.8.11
squash 3d537bc Django 1.8: avoid PIP PEP 440 compatible notation
[...]
pick 93cef62 Django 1.8.x: replace dj.markup with markdown_deux
squash->pick (I think it's ok to keep) 4263018 Django 1.8: remove markdown_deux
[...]
pick 5068d53 Dj1.8: repl longerusername with custom user model
squash->pick 3e00840 Dj18: move CustomUM registration into acc/admin.py
squash f25cc1a Dj1.8: CustomUM: accounts.User: properly swappable
squash 36acce1 Dj1.8: custom user model: update imports
[...]
```

**Comment by vladimir-mencl-eresearch** [comment]
The intention was to `squash` them into the last commit `pick`-ed - e.g., in case of documentation, merge all changes into the commit adding the documentation.

I'm almost done with the squashing - will push soon.

**Comment by vladimir-mencl-eresearch** [comment]
Squashing down to 26 commits done - please let me know whether you're happy with the outcome...

Cheers,
Vlad

**Comment by vladimir-mencl-eresearch** [comment]
For reference, original unsquashed (and un-reordered) branch is archived at https://github.com/REANNZ/djnro/compare/django18-backup-before-squash-2016-03-17

**Comment by zmousm** [comment]
Great work, thanks a lot, merging at last!

**Pull request merged**